Fix part of #14033: Type annotate opportunity domain

megamayoy opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #14033 
2. This PR does the following: add type annotations to `core/domain/opportunity_domain.py ` and  `core/domain/opportunity_domain_test.py `

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
![sc](https://user-images.githubusercontent.com/11338579/138511670-7f6d6301-7969-434a-b147-73803c865950.png)

<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

megamayoy committed:
remove tests which check for passing wrong argument types based on #13528

megamayoy committed:
type annotate opportunity domain files

- add type annotations to opportunity_domain.py and
opportunity domain_test.py

megamayoy committed:
fix linting

megamayoy commented on core/domain/opportunity_domain_test.py:
```diff
@@ -96,96 +96,23 @@ def test_to_and_from_dict_works_correctly(self):
             'translation_in_review_counts': {}
         })
 
-    def test_invalid_topic_id_fails_validation_check(self):
```
Since mypy checks for wrong argument types, tests that check for wrong argument types are removed in this PR.
This is based on this issue https://github.com/oppia/oppia/issues/13528

megamayoy committed:
remove validation against wrong argument types

this is done because mypy does argument type checking so including
this in test is no longer needed.

megamayoy committed:
remove unneeded module

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -19,22 +19,62 @@
 from core.constants import constants
 
+from typing import Dict, List
+import typing_extensions
```
```suggestion
from typing_extensions import TypedDict
```

megamayoy replied:
@vojtechjelinek  When I do this, Lint Checks Manager throws an error 
`C003: Import "TypedDict" from "typing_extensions" is not a module`

megamayoy replied:
Done

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -19,22 +19,62 @@
+    typing_extensions.TypedDict, total=False
+):
+    """A dictionary representing ExplorationOpportunitySummary object.
+    """
```
```suggestion
    """A dictionary representing ExplorationOpportunitySummary object."""
```

megamayoy replied:
Done

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -19,22 +19,62 @@
+):
+    """A dictionary representing ExplorationOpportunitySummary object.
+    """
+
```
```suggestion
```

megamayoy replied:
Done

vojtechjelinek replied:
Not done.

megamayoy replied:
When I delete the newline between the doctring and the rest of the class, the Linting Checks manager raises an error` W9030: Missing single newline below class docstring.`

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -19,22 +19,62 @@
+
+
+class ExplorationOpportunitySummaryDict(
+    typing_extensions.TypedDict, total=False
```
Why `total=False` (I know what it does, but not sure why is it used here.)

megamayoy replied:
That's because the `to_dict()` method doesn't return all the keys of the TypedDict so in order to tell mypy that
it's ok for some keys to be left out , the `total=False` is used

vojtechjelinek replied:
Which keys it doesn't return? Also, where is the full dictionary used.

megamayoy replied:
the keys that is not returned [here](https://github.com/oppia/oppia/blob/63f575f48f5b970eeda5ea95d41c4c062ad094d8/core/domain/opportunity_domain.py#L162) are 
`topic_id`
`story_id`
`incomplete_translation_language_codes`
`language_codes_needing_voice_artists`
`language_codes_with_assigned_voice_artists`
`language_codes_with_assigned_voice_artists`
  



the full dictionary is used [here](https://github.com/oppia/oppia/blob/63f575f48f5b970eeda5ea95d41c4c062ad094d8/core/domain/opportunity_domain.py#L120). the `from_dict()` method expects to receive all keys.

vojtechjelinek replied:
Hmm, then we probably should have two different types, one can inherit from the other.

megamayoy replied:
I think we should create two different types without inheritance because for example 

```
class MovieBase(TypedDict):
    name: str
    year: int

class Movie(MovieBase, total=False):
    based_on: str
```

the `Movie` class will have the `name` and `year` fields as mandatory fields and `based_on` field will be optional
So we can't make required fields in the parent class optional in the child class, that's why we can't use inheritance in our case. 
This is according to the [documentation](https://mypy.readthedocs.io/en/stable/more_types.html#mixing-required-and-non-required-items)

vojtechjelinek replied:
I mean that shouldn't be an issue in our case right? We can do it like this, (not sure abut the naming):
```python
class PartialExplorationOpportunitySummaryDict(TypedDict):
    id: str
    topic_name: str
    story_title: str
    chapter_title: str
    content_count: int
    translation_counts: Dict[str, int]
    translation_in_review_counts: Dict[str, int]

class ExplorationOpportunitySummaryDict(
    PartialExplorationOpportunitySummaryDict
):
    topic_id: str
    story_id: str
    incomplete_translation_language_codes: List[str]
    language_codes_needing_voice_artists: List[str]
    language_codes_with_assigned_voice_artists: List[str]
```

megamayoy replied:
Yeah, right.
The naming is good to me. It' readable and clear.
I was thinking  of going with `Minimal` instead of `Partial` but I think both are fine.

```
class MininalExplorationOpportunitySummaryDict(TypedDict):
    id: str
    topic_name: str
    story_title: str
    chapter_title: str
    content_count: int
    translation_counts: Dict[str, int]
    translation_in_review_counts: Dict[str, int]

class ExplorationOpportunitySummaryDict(
    MininalExplorationOpportunitySummaryDict
):
    topic_id: str
    story_id: str
    incomplete_translation_language_codes: List[str]
    language_codes_needing_voice_artists: List[str]
    language_codes_with_assigned_voice_artists: List[str]
```

I'm gonna go with partial

megamayoy replied:
another suggestion for the naming is 

parent class : `ExplorationOpportunitySummaryRequiredFields` as there's a [comment](https://github.com/oppia/oppia/blob/a68bc5ac54d4f6817360ddefa3b0c1ca70b09639/core/domain/opportunity_domain.py#L159) that says that these are the required fields to represent an opportunity to a contributor.

child class :  `ExplorationOpportunitySummaryDict`

megamayoy replied:
Done.
Let me know if you prefer a better naming.

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -19,22 +19,62 @@
+    """A dictionary representing SkillOpportunity object.
+    """
+
+    id: str
```
```suggestion
    """A dictionary representing SkillOpportunity object."""
    id: str
```

megamayoy replied:
Done

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -193,7 +214,10 @@ def validate(self):
-    def _validate_translation_counts(self, translation_counts):
+    def _validate_translation_counts(
+        self,
+        translation_counts: Dict[str, int]
```
```suggestion
        self, translation_counts: Dict[str, int]
```

megamayoy replied:
Done

vojtechjelinek commented on core/domain/opportunity_domain_test.py:
```diff
@@ -64,21 +63,22 @@ def setUp(self):
-            'language_codes_needing_voice_artists': ['en'],
-            'language_codes_with_assigned_voice_artists': [],
-            'translation_in_review_counts': {}
-        }
```
Keep this as is only set it's type to `ExplorationOpportunitySummaryDict`

megamayoy replied:
Done

vojtechjelinek commented on core/domain/opportunity_domain_test.py:
```diff
@@ -371,17 +291,18 @@ def test_all_languages_in_summary_equals_supported_languages(self):
+                opportunity_domain.SkillOpportunityDict(
+                    id='skill_1',
+                    skill_description='A new skill',
+                    question_count=10
```
Again keep normal dict just set its type.

megamayoy replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

megamayoy committed:
fix linting

megamayoy commented:
@vojtechjelinek All requested changes are done except for changing the import statement `import typing_extensions`
to `from typing_extensions import TypedDict`. Linting checks throws an error
`C003: Import "TypedDict" from "typing_extensions" is not a module`

vojtechjelinek commented:
> @vojtechjelinek All requested changes are done except for changing the import statement `import typing_extensions` to `from typing_extensions import TypedDict`. Linting checks throws an error `C003: Import "TypedDict" from "typing_extensions" is not a module`

You will need to add `'typing_extensions'` to https://github.com/oppia/oppia/blob/develop/scripts/linters/pylint_extensions.py#L1372

vojtechjelinek commented:
@megamayoy Please, reply to all comments that you have fixed with "Done."

megamayoy committed:
Allow to directly import TypedDict from typing_extensions module

megamayoy committed:
Create a partial TypeDict for ExplorationOpportunitySummary objects

This is done so that mypy checks for the partial fields
that are required in the to_dict() method

megamayoy commented:
> > @vojtechjelinek All requested changes are done except for changing the import statement `import typing_extensions` to `from typing_extensions import TypedDict`. Linting checks throws an error `C003: Import "TypedDict" from "typing_extensions" is not a module`
> 
> You will need to add `'typing_extensions'` to https://github.com/oppia/oppia/blob/develop/scripts/linters/pylint_extensions.py#L1372

Done

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -19,22 +19,67 @@
+class PartialExplorationOpportunitySummaryDict(TypedDict):
+    """A dictionary representing partial fields of
+    ExplorationOpportunitySummary object.
+    """
```
Please explain here when this partial dict is used.

megamayoy replied:
Done. Let me know if the edited doctring needs more explanation

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -19,22 +19,67 @@
+
+
+class SkillOpportunityDict(TypedDict):
+    """A dictionary representing SkillOpportunity object."""
+
+    id: str
```
```suggestion
    """A dictionary representing SkillOpportunity object."""
    id: str
```

megamayoy replied:
When I delete the newline between the doctring and the rest of the class, the Linting Checks manager raises an error `W9030: Missing single newline below class docstring.`

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -19,22 +19,67 @@
+):
+    """A dictionary representing ExplorationOpportunitySummary object."""
+
+    topic_id: str
```
```suggestion
    """A dictionary representing ExplorationOpportunitySummary object."""
    topic_id: str
```

Please try to explain when this dict is used and how it differs from the partial dict, in the docstring.

megamayoy replied:
Done. I'm not sure if I clarified it correctly. I only wrote that this dict gets the required fields to represent 
an opportunity from **PartialExplorationOpportunitySummaryDict**.

megamayoy replied:
@vojtechjelinek the same thing happens here When I delete the newline between the doctring and the rest of the class, the Linting Checks manager raises an error` W9030: Missing single newline below class docstring.`

vojtechjelinek commented on core/domain/opportunity_domain_test.py:
```diff
@@ -64,20 +63,22 @@ def setUp(self):
+        'translation_counts': {},
+        'language_codes_needing_voice_artists': ['en'],
+        'language_codes_with_assigned_voice_artists': [],
+        'translation_in_review_counts': {}
```
```suggestion
            'id': 'exp_1',
            'topic_id': 'topic_1',
            'topic_name': 'A topic',
            'story_id': 'story_1',
            'story_title': 'A new story',
            'chapter_title': 'A new chapter',
            'content_count': 5,
            'incomplete_translation_language_codes': ['hi-en', 'hi'],
            'translation_counts': {},
            'language_codes_needing_voice_artists': ['en'],
            'language_codes_with_assigned_voice_artists': [],
            'translation_in_review_counts': {}
```

megamayoy replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

megamayoy committed:
fix indentation

BenHenning commented on CONTRIBUTORS:
```diff
@@ -247,6 +247,7 @@ Luis Ulloa <scssaluis@gmail.com>
 Lunrong Chen <lunrong@google.com>
 Madhav Sainanee <madhavsainanee123@gmail.com>
 Madiyar Aitbayev <Madiyar92@gmail.com>
+Mai Elshiashi <mai.elshiashi@gmail.com>
```
Ah, you don't need to update this in your PR @megamayoy. For one, we should keep PRs specific to the change they relate to, and second, we update the attribution files when we prepare for a release. Since we haven't yet shipped a new version of Oppia since you started contributing code, your name hasn't yet been added.

Please remove this and do let us know if you aren't included after we release the next version of Oppia.

Thanks!

megamayoy replied:
Done

BenHenning replied:
Thanks!

BenHenning reviewed (dismissed):
Just had one comment per my codeowner files. PTAL.

megamayoy committed:
remove name from CONTRIBUTORS file

BenHenning reviewed (commented):
Thanks @megamayoy. Confirming that my codeowners is no longer needed on this PR, so removing myself. If you do need me to take a review, please re-add me.

hardikkat24 commented on core/domain/opportunity_domain.py:
```diff
@@ -128,34 +176,12 @@ def to_dict(self):
-        if not isinstance(self.content_count, int):
-            raise utils.ValidationError(
-                'Expected content_count to be an integer, received %s' %
-                self.content_count)
```
@vojtechjelinek Are we sure that we want to remove type validations and depend on mypy for this? Thanks!

vojtechjelinek replied:
I think type validations can be removed, I see how they can still catch some errors, but I think it is only a tiny subset of errors that can go through this.

hardikkat24 replied:
Agreed. Thanks!

hardikkat24 reviewed (changes requested):
Type annotations look good. I'm unsure whether we should remove type validation and depend on mypy for it. Thanks!

megamayoy committed:
Merge branch 'develop' of github.com:oppia/oppia into type-annotate-opportunity-domain

megamayoy committed:
Edit docstring

megamayoy committed:
edit a doctring

megamayoy commented:
@vojtechjelinek could you please restart the tests ?

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -22,18 +22,71 @@
+    """A dictionary representing ExplorationOpportunitySummary object.
+
+    contains all fields of an ExplorationOpportunitySummary object
+    It gets the required fields from PartialExplorationOpportunitySummaryDict
```
```suggestion
    Contains all fields of an ExplorationOpportunitySummary object.
    It gets the required fields from PartialExplorationOpportunitySummaryDict.
```

megamayoy replied:
Done

vojtechjelinek commented on core/domain/opportunity_domain.py:
```diff
@@ -22,18 +22,71 @@
+    ExplorationOpportunitySummary object.
+
+    This dict has only required fields to represent
+    an opportunity to a contributor
```
```suggestion
    This dict has only required fields to represent
    an opportunity to a contributor.
```

megamayoy replied:
Done

vojtechjelinek reviewed (approved):
LGTM! Few nits.

megamayoy committed:
edit docstrings

DubeySandeep reviewed (approved):
LGTM for the changes in `scripts/linters/pylint_extensions.py`

megamayoy commented:
@vojtechjelinek As this is the last day of the Hacktobefest and all requested changes are done, I'd really appreciate it if you add the `hacktoberfest-accepted` label on this PR if it's not gonna be merged by the end of the day. It's been a pleasure contributing to Oppia and definitely won't be the last. Thank you!

vojtechjelinek reviewed (approved):
LGTM!

hardikkat24 reviewed (approved):
LGTM. Thanks @megamayoy!

sagangwee commented on core/domain/opportunity_domain.py:
```diff
@@ -22,18 +22,71 @@
+from typing_extensions import TypedDict
+
+
+class PartialExplorationOpportunitySummaryDict(TypedDict):
```
Maybe `ExplorationOpportunitySummaryCard` to emphasize that this is the minimum data needed to represent the opportunity on the frontend.

megamayoy replied:
@sagangwee  I'm not sure about the `Card` suffix. This class represents a data type not an actual card in the system, right? I'm afraid readers of the code are gonna confuse between domain names and types names if we added the suffix "Card". How about going with a name that this dictionary really represents .Something like `ExplorationOpportunitySummaryRequiredFields` I know it maybe too long but it expresses exactly what this dictionary is representing

sagangwee reviewed (approved):
@megamayoy LGTM, minus one naming comment. Thanks!

sajalasati reviewed (changes requested):
Thanks @megamayoy, just one thing - for both the files can you replace `absolute_import` and `unicode_literals` future imports with `annotations` future import, and remove quotes from the type annotation. You can take a look at [this PR](https://github.com/oppia/oppia/pull/14132/files#diff-525d9ef50f6a3c3a4b1676d564e8d2ca037ae396db1e88689c438593fcecd313) as an example.

megamayoy committed:
Merge branch 'develop' of github.com:oppia/oppia into type-annotate-opportunity-domain

megamayoy commented:
> Thanks @megamayoy, just one thing - for both the files can you replace `absolute_import` and `unicode_literals` future imports with `annotations` future import, and remove quotes from the type annotation. You can take a look at [this PR](https://github.com/oppia/oppia/pull/14132/files#diff-525d9ef50f6a3c3a4b1676d564e8d2ca037ae396db1e88689c438593fcecd313) as an example.

@sajalasati  I think this change is done already by @vojtechjelinek

sajalasati reviewed (approved):
LGTM!

sajalasati commented:
@megamayoy Just a suggestion - instead of re-requesting for review, please leave `PTAL @reviewer-name` comment on the thread, or manually assign them (if you can) and unassign yourself so that the reviewer knows its their turn now.
